### ISSUE TITLE
To not allow `[ '' ]` as `staticDirs` or `outputDir` argument in `buildStaticStandalone()` method

### ISSUE THREAD

**Issue by oscard0m** [storybookjs/storybook#13860]
When calling to `copyAllStaticFiles()` with `['']` as argument (by mistake), it executes and ends up copying root folder and more collateral side effects happening.

https://github.com/storybookjs/storybook/blob/59ee9c3e4e4e9b2d2b80dafe503cbfac669df279/lib/core/src/server/build-static.ts#L104

Would be nice to throw, or not allow that value at TS level.

I would be happy to PR a fix if you agree on the problem

**Comment by shilman** [comment]
That sounds like a good thing to check for. A PR is welcome!

**Comment by shilman** [comment]
Shiver me timbers!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.2.0-alpha.32 containing PR #13969 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb upgrade --prerelease
```

**Comment by oscard0m** [comment]
Hi @shilman,

Tried to upgrade in local (`npx sb upgrade --prerelease`) with a small snippet of code:

```js
const storybook = require('@storybook/react/standalone');

storybook({
	mode: 'static',
	outputDir: ''
})
```

And I'm getting the following error:

<details>
<summary>Error: Cannot find module './dist/cjs/server/standalone'</summary>

<code>
Error: Cannot find module './dist/cjs/server/standalone'
Require stack:
- /Users/XXXXXXXXX/dev/experiments/storybook/node_modules/@storybook/core/standalone.js
- /Users/XXXXXXXXX/dev/experiments/storybook/node_modules/@storybook/react/standalone.js
- /Users/XXXXXXXXX/dev/experiments/storybook/src/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/Users/XXXXXXXXX/dev/experiments/storybook/node_modules/@storybook/core/standalone.js:1:15)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/XXXXXXXXX/dev/experiments/storybook/node_modules/@storybook/core/standalone.js',
    '/Users/XXXXXXXXX/dev/experiments/storybook/node_modules/@storybook/react/standalone.js',
    '/Users/XXXXXXXXX/dev/experiments/storybook/src/index.js'
  ]
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! storybook-experiment@1.0.0 start: `node ./src/index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the storybook-experiment@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/XXXXXXXXX/.npm/_logs/2021-02-21T12_28_01_668Z-debug.log
</code>

</details>


---

Checking package versions in node modules, seems like the version has been upgraded properly

Any clue on what could be the issue? Should I create a bug and research on it?

**Comment by shilman** [comment]
@oscard0m The other PR broke it. I'll should be able to put out a fix today. Thanks for the heads up! The command line should still work

**Comment by oscard0m** [comment]
# ⚠️ !!!!

I just noticed the bug I solved does not match with the one I reported in the issue. I'm going to solve the one described in the issue right now. They are very similar:

- one is avoiding an empty string for `staticDir` prop
- the other one (which was in reality solved by https://github.com/storybookjs/storybook/pull/13969) is to avoid an empty for `outputDir` prop

How should I proceed to solve this mess? @shilman 

Once the issue and commits fix in `next` branch is solved I will open a PR for fixing `staticDir` one

**Comment by oscard0m** [comment]
> @oscard0m The other PR broke it. I'll should be able to put out a fix today. Thanks for the heads up! The command line should still work

no problem @shilman 

I'm going to try it through command line. Thanks for the quick reply ❤️

**Comment by oscard0m** [comment]
> > @oscard0m The other PR broke it. I'll should be able to put out a fix today. Thanks for the heads up! The command line should still work
> 
> no problem @shilman
> 
> I'm going to try it through command line. Thanks for the quick reply ❤️

Apparently I'm not able to exactly try my changes due to this line:
https://github.com/storybookjs/storybook/blob/0327cae8b8330b4206158b765d87bbfe43f15b89/lib/core-server/src/build-static.ts#L105

No matter if I pass it through `package.json` or `cli arg`, an empty string `''` will be falsy in this line and will fallback to `./storybook-static`

Checking this code, an idea coming to my mind:
Should this fallback `./storybook/static` added [inside `buildStatcStandalone()`](https://github.com/storybookjs/storybook/blob/0327cae8b8330b4206158b765d87bbfe43f15b89/lib/core-server/src/build-static.ts#L29-L31) instead?

```diff
- if (options.outputDir === '') {
+ if (!options.outputDir) {
-    throw new Error("Won't remove current directory. Check your outputDir!");
+    options.outputDir = './storybook-static' 
}
```

**Comment by shilman** [comment]
```diff
- if (options.outputDir === '') {
+ if (!options.outputDir) {
-    throw new Error("Won't remove current directory. Check your outputDir!");
+    options.outputDir = './storybook-static' 
}
```

@oscard0m 👍

**Comment by oscard0m** [comment]
Just a followup on this @shilman (I guess you are busy, no rush), I would just like to avoid releasing the code is in next regarding this right now.

**Comment by shilman** [comment]
@oscard0m does the test from this PR pass even without the change? if so, create a new PR that reverts that change (but not the test) and also adds the new test/logic. WDYT?

**Comment by oscard0m** [comment]
> @oscard0m does the test from this PR pass even without the change? if so, create a new PR that reverts that change (but not the test) and also adds the new test/logic. WDYT? thanks for following up and being thorough on this! 💪

# TLDR: 
- 🍾  No need to revert.
- Edited title of the issue to cover both cases:
   - ✅ `outputDir` (https://github.com/storybookjs/storybook/pull/13969)
   - 🏗 `staticDir` (WIP)
- There will be a refactor proposal in a different pull request

---

@shilman this it what happens:

[`buildStaticStandalone()`](https://github.com/oscard0m/storybook/blob/4ecaf7c7644e5463624818aece7ecc258c73a58e/lib/core-server/src/build-static.ts#L25) method, if it is called explicitly (as we do in the test [I added](https://github.com/oscard0m/storybook/blob/0c68ffe182140cd8180e169e3c594dd020b329bb/lib/core-server/src/core-presets.test.ts#L201)), it deletes the current folder without the fix introduced in [this PR](https://github.com/storybookjs/storybook/pull/13969) and throws an error with that PR. 

**This  is good 👍🏽** 

What I noticed is the only place where `buildStaticStandalone()` is invoked in [here](https://github.com/oscard0m/storybook/blob/4ecaf7c7644e5463624818aece7ecc258c73a58e/lib/core-server/src/standalone.ts#L21), and it [always fallbacks to a controlled path](https://github.com/oscard0m/storybook/blob/4ecaf7c7644e5463624818aece7ecc258c73a58e/lib/core-server/src/build-static.ts#L105), so no chance to pass an empty string. So... in the end what we merged is not that bad. What I'm going to do is to suggest a new refactor on top of the already merged PR.

**Comment by shilman** [comment]
Huzzah!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.2.0-rc.3 containing PR #14068 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb upgrade --prerelease
```

Closing this issue. Please re-open if you think there's still more to do.

**Comment by oscard0m** [comment]
> * There will be a refactor proposal in a different pull request

I will wait to this https://github.com/storybookjs/storybook/issues/13250 because probably will close the need I had in mind
